core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +96,195 @@ pub struct ShardLayoutV1 {
     /// Maps shards from the last shard layout to shards that it splits to in this shard layout,
     /// Useful for constructing states for the shards.
     /// None for the genesis shard layout
-    shards_split_map: Option<ShardSplitMap>,
+    shards_split_map: Option<ShardsSplitMap>,
     /// Maps shard in this shard layout to their parent shard
     /// Since shard_ids always range from 0 to num_shards - 1, we use vec instead of a hashmap
     to_parent_shard_map: Option<Vec<ShardId>>,
     /// Version of the shard layout, this is useful for uniquely identify the shard layout
     version: ShardVersion,
 }
 
+impl ShardLayoutV1 {
+    // In this shard layout the accounts are divided into ranges, each range is
+    // mapped to a shard. The shards are contiguous and start from 0.
+    fn account_id_to_shard_id(&self, account_id: &AccountId) -> ShardId {
+        // Note: As we scale up the number of shards we can consider
+        // changing this method to do a binary search rather than linear
+        // scan. For the time being, with only 4 shards, this is perfectly fine.
+        let mut shard_id: ShardId = 0;
+        for boundary_account in &self.boundary_accounts {
+            if account_id < boundary_account {
+                break;
+            }
+            shard_id += 1;
+        }
+        shard_id
+    }
+}
+
+/// A boundary of an account range. Can be the start, the end or an account id
+/// in between two shards.
+#[derive(serde::Serialize, serde::Deserialize, Clone, PartialEq, Eq, Ord)]
+pub enum AccountBoundary {
```

Comment:
maybe give boudaries for a concrete example?, eg. if layout is ["aa", "kk", "tt"], the first boundary is Start-Middle(aa) and the second one is Middle(aa)-Middle(kk) etc. It did not immediately occurred to me that `Middle(AccountId)` are the borders instead of accounts in the middle.

Reply:
Will do, if you have better naming suggestions I'm open.